src/java/org/apache/cassandra/cql3/functions/UDFContextImpl.java

```diff
@@ -28,26 +28,29 @@
 import org.apache.cassandra.schema.CQLTypeParser;
 import org.apache.cassandra.schema.KeyspaceMetadata;
 import org.apache.cassandra.cql3.functions.types.*;
+import org.apache.cassandra.schema.Schema;
 import org.apache.cassandra.utils.ByteBufferUtil;
 
 /**
  * Package private implementation of {@link UDFContext}
  */
 public final class UDFContextImpl implements UDFContext
 {
-    private final KeyspaceMetadata keyspaceMetadata;
+    private final String keyspace;
     private final Map<String, TypeCodec<Object>> byName = new HashMap<>();
     private final TypeCodec<Object>[] argCodecs;
     private final TypeCodec<Object> returnCodec;
 
+    private volatile KeyspaceMetadata metadata = null;
```

Comment:
I'd annotate this with `@Nullable`. Or, even better, the entire `getMetadata()` method can be gone if we replace this by a memoizing supplier:
```java
/** The keyspace metadata could not be available at startup, see CASSANDRA-18739 */
private final Supplier<KeyspaceMetadata> metadata;

UDFContextImpl(List<ColumnIdentifier> argNames, TypeCodec<Object>[] argCodecs, TypeCodec<Object> returnCodec,
               String keyspace)
{
    ...
    metadata = Suppliers.memoize(() -> Schema.instance.getKeyspaceMetadata(keyspace));
}
```

Reply:
good idea